File: state/java/State.java
// State interface
interface DocumentState {
    String publish(Document doc);
    String review(Document doc);
}

// Context
class Document {
    private DocumentState state;
    private String name;

    public Document(String name) {
        this.name = name;
        this.state = new DraftState();
    }

    public void setState(DocumentState state) {
        this.state = state;
    }

    public String publish() {
        return state.publish(this);
    }

    public String review() {
        return state.review(this);
    }
}

// Concrete States
class DraftState implements DocumentState {
    @Override
    public String publish(Document doc) {
        return "Cannot publish from Draft. Please review first.";
    }

    @Override
    public String review(Document doc) {
        doc.setState(new ReviewState());
        return "Document moved to Review state";
    }
}

class ReviewState implements DocumentState {
    @Override
    public String publish(Document doc) {
        doc.setState(new PublishedState());
        return "Document published successfully";
    }

    @Override
    public String review(Document doc) {
        return "Document is already in Review state";
    }
}

class PublishedState implements DocumentState {
    @Override
    public String publish(Document doc) {
        return "Document is already published";
    }

    @Override
    public String review(Document doc) {
        return "Cannot review a published document";
    }
}

// Main class demonstrating State pattern
public class State {
    public static void main(String[] args) {
        Document doc = new Document("Design Patterns Guide");

        System.out.println("Initial state: Draft");
        System.out.println(doc.publish());
        System.out.println(doc.review());
        System.out.println(doc.publish());
    }
}

File: state/typescript/State.ts
// State interface
interface DocumentState {
  publish(doc: Document): string;
  review(doc: Document): string;
}

// Context
class Document {
  private state: DocumentState;
  private name: string;

  constructor(name: string) {
    this.name = name;
    this.state = new DraftState();
  }

  setState(state: DocumentState): void {
    this.state = state;
  }

  publish(): string {
    return this.state.publish(this);
  }

  review(): string {
    return this.state.review(this);
  }
}

// Concrete States
class DraftState implements DocumentState {
  publish(doc: Document): string {
    return 'Cannot publish from Draft. Please review first.';
  }

  review(doc: Document): string {
    doc.setState(new ReviewState());
    return 'Document moved to Review state';
  }
}

class ReviewState implements DocumentState {
  publish(doc: Document): string {
    doc.setState(new PublishedState());
    return 'Document published successfully';
  }

  review(doc: Document): string {
    return 'Document is already in Review state';
  }
}

class PublishedState implements DocumentState {
  publish(doc: Document): string {
    return 'Document is already published';
  }

  review(doc: Document): string {
    return 'Cannot review a published document';
  }
}

// Usage
const doc = new Document('Design Patterns Guide');

console.log('Initial state: Draft');
console.log(doc.publish());
console.log(doc.review());
console.log(doc.publish());
